todos first project

AmitSilber opened the pull request:
todos: initial commit

AmitSilber committed:
Style.css: initial commit.

AmitSilber committed:
index.html: initial commit

AmitSilber committed:
main.js: initial commit

mfocshaner commented on scripts/main.js:
```diff
@@ -0,0 +1,191 @@
+    else
+        currentRow.before(row);
+}
+function init() {
+    if (localStorage.length === 0) {
+        localStorage.setItem("length", "0");
+        createNewItem("Hi! Im a new task on your todo list", "03-05-2023")
+        return;
+    }
+    const length = localStorage.getItem("length");
+    for (let key = 0; key < length; key++) {
+        renderItem(JSON.parse(localStorage.getItem(key.toString())));
+    
+    }
+    
+}
```
As discussed, split logic; maybe extract this to a function called "loadItemsFromStorage"

mfocshaner commented on scripts/main.js:
```diff
@@ -0,0 +1,191 @@
+ * render (or reload) new item to table
+ * @param {title,date,status,index} itemObj 
+ */
+function renderItem(itemObj) {
+    const row = todoTable.insertRow(-1);
+
+    const titleEdditButtonsCell = row.insertCell(0); // title + edit + delete buttons 
+    const dueDateCell = row.insertCell(1); // date 
+    const statusCell = row.insertCell(2); // status
+    const titleEditToggleCell = document.createElement("div") // title + edit wrapper
+    const editTitleModeCell = document.createElement("input"); // edit title slot
+    const titleCell = document.createElement("p");
+    const deleteCell = document.createElement("span"); 
+    const editCell = document.createElement("span");
+
+    dueDateCell.draggable = "true"; // enable row dragging by date column
+    editTitleModeCell.id = "edit-title" 
+    editTitleModeCell.style.setProperty("display", "none"); // make edit title invisible by default
+    row.setAttribute("key", itemObj.index);
+    titleCell.id = "title-text"
+    statusCell.innerHTML = itemObj.status ? checkIcon: "";
+    titleCell.innerHTML = itemObj.title;
+    dueDateCell.innerHTML = itemObj.date;
+    editTitleModeCell.value = itemObj.title;
+    deleteCell.textContent = deleteIcon;
+    editCell.textContent = editIcon
+    statusCell.className = "hover-cell"
+    editCell.className = "hover-cell"
+    deleteCell.className = "hover-cell"
+
```
this can be split into smaller functions, i.e. `renderDeleteCell`

mfocshaner commented on scripts/main.js:
```diff
@@ -0,0 +1,191 @@
+ * @param {title,date,status,index} itemObj 
+ */
+function renderItem(itemObj) {
+    const row = todoTable.insertRow(-1);
```
why do you do this?

mfocshaner commented on scripts/main.js:
```diff
@@ -0,0 +1,191 @@
+var row;
+
+/**
+ * render (or reload) new item to table
```
is this only for new items? if so the function can be named "renderNewItem"

mfocshaner commented on scripts/main.js:
```diff
@@ -0,0 +1,191 @@
+    const key = row.getAttribute("key")
+    let length = localStorage.getItem("length");
+    let next = row
+    localStorage.removeItem(key);
+    for (let i = Number(key) + 1; i < length; i++) { // undate succesive rows
+        next = next.nextElementSibling;
+        updateItemIndex(i, next);
+    }
+    length--;
+    localStorage.setItem("length", length);
```
you save each object in the local storage in its own key, maybe it will be easier for you to save 1 entry which is an array of objects?

AmitSilber replied:
just thought that loading and downloading the entire todo list for every individual read/write would be less efficient.

mfocshaner reviewed (commented):
Looks good, see comments, and lets do a round

AmitSilber committed:
- changed localStorage structure from key value pairs to list
- split render into different cell renders
- added drag end to update order of list after drag is over
- split init into two different inits
- organized order of functions
